File: sdk/node/cookies.js
// Explicit, authenticated Cookie migration. This does not change OSCrypt or read
// another profile's database. Only the supplied live Chromium context is used.
import { createCipheriv, createDecipheriv, randomBytes, scrypt as scryptCallback } from "node:crypto";
import { promisify } from "node:util";
import { constants } from "node:fs";
import { link, lstat, open, unlink } from "node:fs/promises";
import { dirname, basename, join } from "node:path";
import { isIP } from "node:net";

const scrypt = promisify(scryptCallback);
const MAGIC = Buffer.from("CHROMIX-COOKIES\0\x01", "binary");
const HEADER = MAGIC.length + 16 + 12;
const MAX_BYTES = 16 * 1024 * 1024;
const MAX_COOKIES = 10000;
const fail = () => new Error("Invalid or unsupported portable Cookie data");

function passwordBytes(password) {
  if (typeof password !== "string" || /[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?<![\uD800-\uDBFF])[\uDC00-\uDFFF]/u.test(password))
    throw new TypeError("passphrase must be valid Unicode text");
  const bytes = Buffer.from(password, "utf8");
  if (bytes.length < 12 || bytes.length > 1024)
    throw new TypeError("passphrase must contain 12–1024 UTF-8 bytes");
  return bytes;
}

function text(value, maximum, allowEmpty = true) {
  if (typeof value !== "string" || (!allowEmpty && !value) || Buffer.byteLength(value) > maximum ||
      /[\x00-\x1f\x7f]/.test(value) || Buffer.from(value).toString("utf8") !== value) throw fail();
  return value;
}

function hostForDomain(domain) {
  const host = domain.startsWith(".") ? domain.slice(1) : domain;
  const bare = host.startsWith("[") && host.endsWith("]") ? host.slice(1, -1) : host;
  if (isIP(bare) === 6 && !bare.includes("%") && !domain.startsWith(".")) return `[${bare}]`;
  if (!/^[a-z0-9_-]+(?:\.[a-z0-9_-]+)*\.?$/i.test(host) || host.replace(/\.$/, "").length > 253) throw fail();
  return host;
}

export function normalizeCookies(cookies) {
  if (!Array.isArray(cookies) || cookies.length > MAX_COOKIES) throw fail();
  const keys = new Set();
  return cookies.map(input => {
    if (!input || typeof input !== "object" || Array.isArray(input)) throw fail();
    const cookie = {
      name: text(input.name, 4096), value: text(input.value, 16384),
      domain: text(input.domain, 255, false), path: text(input.path, 4096, false),
      secure: input.secure, httpOnly: input.httpOnly,
    };
    hostForDomain(cookie.domain);
    if (!cookie.path.startsWith("/") || typeof cookie.secure !== "boolean" || typeof cookie.httpOnly !== "boolean" ||
        /[;=]/.test(cookie.name) || /;/.test(cookie.value)) throw fail();
    if (input.partitionKeyOpaque === true) throw new Error("Opaque partition Cookies cannot be migrated");
    for (const [key, choices] of [["sameSite", ["Strict", "Lax", "None"]],
      ["priority", ["Low", "Medium", "High"]], ["sourceScheme", ["Unset", "NonSecure", "Secure"]]]) {
      if (input[key] !== undefined) {
        if (!choices.includes(input[key])) throw fail();
        cookie[key] = input[key];
      }
    }
    if (input.expires !== undefined) {
      if (typeof input.expires !== "number" || !Number.isFinite(input.expires) ||
          (input.expires !== -1 && input.expires <= 0)) throw fail();
      if (input.expires !== -1) cookie.expires = input.expires;
    }
    if (input.sourcePort !== undefined) {
      if (!Number.isInteger(input.sourcePort) || input.sourcePort < -1 || input.sourcePort > 65535) throw fail();
      cookie.sourcePort = input.sourcePort;
    }
    if (input.partitionKey !== undefined) {
      const partition = input.partitionKey;
      if (!partition || typeof partition !== "object" || Array.isArray(partition) ||
          typeof partition.hasCrossSiteAncestor !== "boolean") throw fail();
      const site = text(partition.topLevelSite, 2048, false);
      let url;
      try { url = new URL(site); } catch { throw fail(); }
      if (!["http:", "https:"].includes(url.protocol) || url.username || url.password || url.origin !== site) throw fail();
      hostForDomain(url.hostname);
      cookie.partitionKey = { topLevelSite: site, hasCrossSiteAncestor: partition.hasCrossSiteAncestor };
    }
    if (cookie.name.startsWith("__Secure-") && !cookie.secure) throw fail();
    if (cookie.name.startsWith("__Host-") && (!cookie.secure || cookie.path !== "/" || cookie.domain.startsWith("."))) throw fail();
    if ((cookie.sameSite === "None" || cookie.partitionKey) && !cookie.secure) throw fail();
    const key = JSON.stringify([cookie.name, cookie.domain, cookie.path, cookie.partitionKey ?? null]);
    if (keys.has(key)) throw new Error("Duplicate portable Cookie identity");
    keys.add(key);
    return cookie;
  });
}

export async function encryptCookies(cookies, passphrase) {
  const payload = Buffer.from(JSON.stringify({ format: "chromix.cookies", version: 1, cookies: normalizeCookies(cookies) }));
  if (payload.length > MAX_BYTES - HEADER - 16) throw fail();
  const password = passwordBytes(passphrase);
  const salt = randomBytes(16), nonce = randomBytes(12);
  const header = Buffer.concat([MAGIC, salt, nonce]);
  const key = await scrypt(password, salt, 32, { N: 32768, r: 8, p: 1, maxmem: 64 * 1024 * 1024 });
  try {
    const cipher = createCipheriv("aes-256-gcm", key, nonce);
    cipher.setAAD(header);
    return Buffer.concat([header, cipher.update(payload), cipher.final(), cipher.getAuthTag()]);
  } finally { key.fill(0); password.fill(0); payload.fill(0); }
}

export async function decryptCookies(data, passphrase) {
  if (!(data instanceof Uint8Array)) throw fail();
  data = Buffer.from(data);
  if (data.length < HEADER + 16 || data.length > MAX_BYTES || !data.subarray(0, MAGIC.length).equals(MAGIC)) throw fail();
  const password = passwordBytes(passphrase);
  const salt = data.subarray(MAGIC.length, MAGIC.length + 16), nonce = data.subarray(MAGIC.length + 16, HEADER);
  const key = await scrypt(password, salt, 32, { N: 32768, r: 8, p: 1, maxmem: 64 * 1024 * 1024 });
  let payload;
  try {
    const decipher = createDecipheriv("aes-256-gcm", key, nonce);
    decipher.setAAD(data.subarray(0, HEADER));
    decipher.setAuthTag(data.subarray(-16));
    payload = Buffer.concat([decipher.update(data.subarray(HEADER, -16)), decipher.final()]);
    const value = JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(payload));
    if (value?.format !== "chromix.cookies" || value.version !== 1) throw fail();
    return normalizeCookies(value.cookies);
  } catch { throw new Error("Cannot decrypt portable Cookies: wrong passphrase, damaged data or unsupported payload"); }
  finally { key.fill(0); password.fill(0); payload?.fill(0); }
}

async function readEnvelope(path) {
  if (!(await lstat(path)).isFile()) throw fail();
  const stream = await open(path, constants.O_RDONLY | (constants.O_NOFOLLOW ?? 0));
  try {
    const stat = await stream.stat();
    if (!stat.isFile() || stat.size > MAX_BYTES) throw fail();
    const parts = []; let total = 0;
    while (true) {
      const chunk = Buffer.alloc(Math.min(65536, MAX_BYTES + 1 - total));
      const { bytesRead } = await stream.read(chunk);
      if (!bytesRead) break;
      total += bytesRead;
      if (total > MAX_BYTES) throw fail();
      parts.push(chunk.subarray(0, bytesRead));
    }
    return Buffer.concat(parts);
  } finally { await stream.close(); }
}

async function publish(path, data) {
  const temporary = join(dirname(path), `.${basename(path)}.${randomBytes(16).toString("hex")}.tmp`);
  const stream = await open(temporary, "wx", 0o600);
  try {
    try { await stream.writeFile(data); await stream.sync(); }
    finally { await stream.close(); }
    await link(temporary, path); // fail rather than overwrite a prior export
  } finally { await unlink(temporary); }
}

async function withSession(context, action) {
  const page = await context.newPage();
  let session;
  try {
    session = typeof context.newCDPSession === "function" ? await context.newCDPSession(page) : await page.createCDPSession();
    const { targetInfo } = await session.send("Target.getTargetInfo");
    if (targetInfo.type !== "page") throw new Error("Cookie migration requires the supplied context's page target");
    // Network's page-target methods use that page's storage partition, including
    // persistent/default contexts. Storage.* without an id can select the OTHER
    // default context, and Storage.* with an id requires a browser target.
    // getAllCookies is deprecated but still supported by our pinned Chromium.
    return await action(session, {});
  } finally {
    try { if (session) await session.detach(); } finally { await page.close(); }
  }
}

function cookieParams(cookie) {
  const result = { ...cookie };
  if (!cookie.domain.startsWith(".")) {
    const scheme = cookie.sourceScheme === "Secure" || (cookie.sourceScheme !== "NonSecure" && cookie.secure) ? "https" : "http";
    const port = cookie.sourcePort > 0 ? `:${cookie.sourcePort}` : "";
    result.url = `${scheme}://${hostForDomain(cookie.domain)}${port}/`;
    delete result.domain; // Preserve host-only rather than widening to subdomains.
  }
  return result;
}

export async function exportCookies(context, path, { passphrase } = {}) {
  passwordBytes(passphrase).fill(0);
  const cookies = await withSession(context, async (session, params) =>
    (await session.send("Network.getAllCookies", params)).cookies);
  const data = await encryptCookies(cookies, passphrase);
  await publish(path, data);
  return { exported: cookies.length, formatVersion: 1 };
}

export async function importCookies(context, path, { passphrase } = {}) {
  const cookies = await decryptCookies(await readEnvelope(path), passphrase);
  const now = Date.now() / 1000;
  const active = cookies.filter(cookie => cookie.expires === undefined || cookie.expires > now);
  return withSession(context, async (session, params) => {
    const before = await session.send("Network.getAllCookies", params);
    if (before.cookies.length) throw new Error("Cookie import requires an empty destination context; existing Cookies are never cleared");
    if (active.length) await session.send("Network.setCookies", { ...params, cookies: active.map(cookieParams) });
    const after = normalizeCookies((await session.send("Network.getAllCookies", params)).cookies);
    const canonical = values => values.map(value => JSON.stringify(value)).sort();
    if (JSON.stringify(canonical(after)) !== JSON.stringify(canonical(active)))
      throw new Error("Cookie import verification failed; discard this destination context (no automatic clearing or rollback)");
    return { imported: active.length, skippedExpired: cookies.length - active.length, formatVersion: 1 };
  });
}

File: sdk/python/chromix/cookies.py
"""Authenticated, explicit migration of a supplied live Chromium context's Cookies.

Format v1 interoperates with the Node SDK. It does not change OSCrypt, read an
offline profile database, or clear/overwrite Cookies in a populated context.
"""
import hashlib
import ipaddress
import json
import math
import os
from pathlib import Path
import re
import secrets
import stat
import time
from urllib.parse import urlsplit

MAGIC = b"CHROMIX-COOKIES\x00\x01"
HEADER = len(MAGIC) + 16 + 12
MAX_BYTES = 16 * 1024 * 1024


def _invalid():
    return ValueError("Invalid or unsupported portable Cookie data")


def _password(value):
    if not isinstance(value, str):
        raise TypeError("passphrase must be valid Unicode text")
    data = value.encode("utf-8", errors="strict")
    if not 12 <= len(data) <= 1024:
        raise ValueError("passphrase must contain 12–1024 UTF-8 bytes")
    return data


def _text(value, maximum, allow_empty=True):
    if not isinstance(value, str) or (not allow_empty and not value) or re.search(r"[\x00-\x1f\x7f]", value):
        raise _invalid()
    if len(value.encode("utf-8", errors="strict")) > maximum:
        raise _invalid()
    return value


def _host(domain):
    host = domain[1:] if domain.startswith(".") else domain
    bare = host[1:-1] if host.startswith("[") and host.endswith("]") else host
    try:
        if ipaddress.ip_address(bare).version == 6 and not domain.startswith(".") and "%" not in bare:
            return "[" + bare + "]"
    except ValueError:
        pass
    if not re.fullmatch(r"[a-zA-Z0-9_-]+(?:\.[a-zA-Z0-9_-]+)*\.?", host) or len(host.rstrip(".")) > 253:
        raise _invalid()
    return host


def normalize_cookies(cookies):
    if not isinstance(cookies, list) or len(cookies) > 10000:
        raise _invalid()
    result, keys = [], set()
    for value in cookies:
        if not isinstance(value, dict):
            raise _invalid()
        cookie = {"name": _text(value.get("name"), 4096), "value": _text(value.get("value"), 16384),
                  "domain": _text(value.get("domain"), 255, False), "path": _text(value.get("path"), 4096, False),
                  "secure": value.get("secure"), "httpOnly": value.get("httpOnly")}
        _host(cookie["domain"])
        if (not cookie["path"].startswith("/") or type(cookie["secure"]) is not bool or
                type(cookie["httpOnly"]) is not bool or re.search(r"[;=]", cookie["name"]) or ";" in cookie["value"]):
            raise _invalid()
        if value.get("partitionKeyOpaque") is True:
            raise ValueError("Opaque partition Cookies cannot be migrated")
        for field, choices in (("sameSite", ("Strict", "Lax", "None")),
                               ("priority", ("Low", "Medium", "High")),
                               ("sourceScheme", ("Unset", "NonSecure", "Secure"))):
            if field in value:
                if value[field] not in choices:
                    raise _invalid()
                cookie[field] = value[field]
        if "expires" in value:
            expires = value["expires"]
            if type(expires) not in (float, int) or not math.isfinite(expires) or (expires != -1 and expires <= 0):
                raise _invalid()
            if expires != -1:
                cookie["expires"] = expires
        if "sourcePort" in value:
            if type(value["sourcePort"]) is not int or not -1 <= value["sourcePort"] <= 65535:
                raise _invalid()
            cookie["sourcePort"] = value["sourcePort"]
        if "partitionKey" in value:
            partition = value["partitionKey"]
            if not isinstance(partition, dict) or type(partition.get("hasCrossSiteAncestor")) is not bool:
                raise _invalid()
            site = _text(partition.get("topLevelSite"), 2048, False)
            url = urlsplit(site)
            if (url.scheme not in ("http", "https") or not url.hostname or url.username or url.password or
                    url.path or url.query or url.fragment or site != url.scheme + "://" + url.netloc):
                raise _invalid()
            # Require a canonical serialized origin as returned by CDP.
            host = _host(url.hostname)
            if ":" in url.hostname:
                host = "[" + str(ipaddress.IPv6Address(url.hostname)) + "]"
            elif re.fullmatch(r"[0-9]+|0[xX][0-9a-fA-F]+", url.hostname.rstrip(".").rsplit(".", 1)[-1]):
                # Reject WHATWG's noncanonical shorthand/hex/octal IPv4 URLs.
                host = str(ipaddress.IPv4Address(url.hostname))
            port = url.port
            origin = url.scheme + "://" + host.lower()
            if port is not None and port != (443 if url.scheme == "https" else 80):
                origin += ":" + str(port)
            if site != origin:
                raise _invalid()
            cookie["partitionKey"] = {"topLevelSite": site, "hasCrossSiteAncestor": partition["hasCrossSiteAncestor"]}
        if cookie["name"].startswith("__Secure-") and not cookie["secure"]:
            raise _invalid()
        if cookie["name"].startswith("__Host-") and (not cookie["secure"] or cookie["path"] != "/" or cookie["domain"].startswith(".")):
            raise _invalid()
        if (cookie.get("sameSite") == "None" or "partitionKey" in cookie) and not cookie["secure"]:
            raise _invalid()
        key = (cookie["name"], cookie["domain"], cookie["path"], json.dumps(cookie.get("partitionKey"), sort_keys=True))
        if key in keys:
            raise ValueError("Duplicate portable Cookie identity")
        keys.add(key)
        result.append(cookie)
    return result


def _aesgcm():
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError as error:
        raise ImportError("Install the Cookie migration extra: pip install 'chromix[cookies]'") from error
    return AESGCM


def _key(password, salt):
    return hashlib.scrypt(password, salt=salt, n=32768, r=8, p=1, dklen=32, maxmem=64 * 1024 * 1024)


def encrypt_cookies(cookies, passphrase):
    payload = json.dumps({"format": "chromix.cookies", "version": 1, "cookies": normalize_cookies(cookies)},
                         ensure_ascii=True, separators=(",", ":"), allow_nan=False).encode("utf-8")
    if len(payload) > MAX_BYTES - HEADER - 16:
        raise _invalid()
    password = _password(passphrase)
    salt, nonce = secrets.token_bytes(16), secrets.token_bytes(12)
    header = MAGIC + salt + nonce
    return header + _aesgcm()(_key(password, salt)).encrypt(nonce, payload, header)


def decrypt_cookies(data, passphrase):
    if not isinstance(data, bytes) or not HEADER + 16 <= len(data) <= MAX_BYTES or not data.startswith(MAGIC):
        raise _invalid()
    password = _password(passphrase)
    salt, nonce = data[len(MAGIC):len(MAGIC) + 16], data[len(MAGIC) + 16:HEADER]
    aes = _aesgcm()(_key(password, salt))
    try:
        payload = json.loads(aes.decrypt(nonce, data[HEADER:], data[:HEADER]).decode("utf-8"))
        if not isinstance(payload, dict) or payload.get("format") != "chromix.cookies" or type(payload.get("version")) is not int or payload["version"] != 1:
            raise _invalid()
        return normalize_cookies(payload["cookies"])
    except Exception as error:
        raise ValueError("Cannot decrypt portable Cookies: wrong passphrase, damaged data or unsupported payload") from error


def _read(path):
    path = Path(path)
    if not stat.S_ISREG(path.lstat().st_mode):
        raise _invalid()
    fd = os.open(path, os.O_RDONLY | getattr(os, "O_BINARY", 0) | getattr(os, "O_NOFOLLOW", 0))
    with os.fdopen(fd, "rb") as stream:
        if not stat.S_ISREG(os.fstat(stream.fileno()).st_mode):
            raise _invalid()
        data = stream.read(MAX_BYTES + 1)
    if len(data) > MAX_BYTES:
        raise _invalid()
    return data


def _publish(path, data):
    path = Path(path)
    temporary = path.with_name("." + path.name + "." + secrets.token_hex(16) + ".tmp")
    fd = os.open(temporary, os.O_WRONLY | os.O_CREAT | os.O_EXCL | getattr(os, "O_BINARY", 0), 0o600)
    try:
        with os.fdopen(fd, "wb") as stream:
            stream.write(data)
            stream.flush()
            os.fsync(stream.fileno())
        os.link(temporary, path)
    finally:
        temporary.unlink()


def _params(cookie):
    result = dict(cookie)
    if not cookie["domain"].startswith("."):
        secure = cookie.get("sourceScheme") == "Secure" or (cookie.get("sourceScheme") != "NonSecure" and cookie["secure"])
        port = ":" + str(cookie["sourcePort"]) if cookie.get("sourcePort", -1) > 0 else ""
        result["url"] = ("https" if secure else "http") + "://" + _host(cookie["domain"]) + port + "/"
        del result["domain"]
    return result


def _context_params(target):
    if target["targetInfo"].get("type") != "page":
        raise ValueError("Cookie migration requires the supplied context's page target")
    # Network methods use this page's storage partition. Storage.* needs a
    # browser target and otherwise may silently select the default context.
    return {}


def _verify(actual, expected):
    # JSON integer/double spellings differ between Node and Python; compare
    # semantic numeric values, not their JSON serialization (1 == 1.0).
    key = lambda value: (value["domain"], value["path"], value["name"], json.dumps(value.get("partitionKey"), sort_keys=True))
    canonical = lambda values: sorted(normalize_cookies(values), key=key)
    if canonical(actual) != canonical(expected):
        raise ValueError("Cookie import verification failed; discard this destination context (no automatic clearing or rollback)")


def export_cookies(context, path, *, passphrase):
    _password(passphrase)
    page, session = context.new_page(), None
    try:
        session = context.new_cdp_session(page)
        params = _context_params(session.send("Target.getTargetInfo"))
        cookies = session.send("Network.getAllCookies", params)["cookies"]
    finally:
        try:
            if session is not None:
                session.detach()
        finally:
            page.close()
    _publish(path, encrypt_cookies(cookies, passphrase))
    return {"exported": len(cookies), "formatVersion": 1}


def import_cookies(context, path, *, passphrase):
    cookies = decrypt_cookies(_read(path), passphrase)
    now = time.time()
    active = [cookie for cookie in cookies if cookie.get("expires", math.inf) > now]
    page, session = context.new_page(), None
    try:
        session = context.new_cdp_session(page)
        params = _context_params(session.send("Target.getTargetInfo"))
        if session.send("Network.getAllCookies", params)["cookies"]:
            raise ValueError("Cookie import requires an empty destination context; existing Cookies are never cleared")
        if active:
            session.send("Network.setCookies", {**params, "cookies": [_params(cookie) for cookie in active]})
        _verify(session.send("Network.getAllCookies", params)["cookies"], active)
        return {"imported": len(active), "skippedExpired": len(cookies) - len(active), "formatVersion": 1}
    finally:
        try:
            if session is not None:
                session.detach()
        finally:
            page.close()


async def export_cookies_async(context, path, *, passphrase):
    import asyncio
    _password(passphrase)
    page, session = await context.new_page(), None
    try:
        session = await context.new_cdp_session(page)
        params = _context_params(await session.send("Target.getTargetInfo"))
        cookies = (await session.send("Network.getAllCookies", params))["cookies"]
    finally:
        try:
            if session is not None:
                await session.detach()
        finally:
            await page.close()
    def save():
        _publish(path, encrypt_cookies(cookies, passphrase))
    await asyncio.get_running_loop().run_in_executor(None, save)
    return {"exported": len(cookies), "formatVersion": 1}


async def import_cookies_async(context, path, *, passphrase):
    import asyncio
    cookies = await asyncio.get_running_loop().run_in_executor(None, lambda: decrypt_cookies(_read(path), passphrase))
    now = time.time()
    active = [cookie for cookie in cookies if cookie.get("expires", math.inf) > now]
    page, session = await context.new_page(), None
    try:
        session = await context.new_cdp_session(page)
        params = _context_params(await session.send("Target.getTargetInfo"))
        if (await session.send("Network.getAllCookies", params))["cookies"]:
            raise ValueError("Cookie import requires an empty destination context; existing Cookies are never cleared")
        if active:
            await session.send("Network.setCookies", {**params, "cookies": [_params(cookie) for cookie in active]})
        _verify((await session.send("Network.getAllCookies", params))["cookies"], active)
        return {"imported": len(active), "skippedExpired": len(cookies) - len(active), "formatVersion": 1}
    finally:
        try:
            if session is not None:
                await session.detach()
        finally:
            await page.close()
